include/crocoddyl/multibody/residuals/frame-rotation.hpp

```diff
@@ -0,0 +1,181 @@
+                    const Eigen::Ref<const VectorXs>& u);
+
+  /**
+   * @brief Compute the derivatives of the frame rotation residual
+   *
+   * @param[in] data  Frame rotation residual data
+   * @param[in] x     State point \f$\mathbf{x}\in\mathbb{R}^{ndx}\f$
+   * @param[in] u     Control input \f$\mathbf{u}\in\mathbb{R}^{nu}\f$
+   */
+  virtual void calcDiff(const boost::shared_ptr<ResidualDataAbstract>& data, const Eigen::Ref<const VectorXs>& x,
+                        const Eigen::Ref<const VectorXs>& u);
+
+  /**
+   * @brief Create the frame rotation residual data
+   */
+  virtual boost::shared_ptr<ResidualDataAbstract> createData(DataCollectorAbstract* const data);
+
+  /**
+   * @brief Return the reference frame id
+   */
+  pinocchio::FrameIndex get_id() const;
+
+  /**
+   * @brief Return the reference frame rotation
+   */
+  const Matrix3s& get_reference() const;
+
+  /**
+   * @brief Modify the reference frame id
+   */
+  void set_id(const pinocchio::FrameIndex id);
+
+  /**
+   * @brief Modify the reference frame rotation
+   */
+  void set_reference(const Matrix3s& reference);
+
+  using Base::nu_;
+  using Base::state_;
+  using Base::u_dependent_;
+  using Base::unone_;
+  using Base::v_dependent_;
+
+ private:
+  pinocchio::FrameIndex id_;                                              //!< Reference frame id
+  Matrix3s Rref_;                                                         //!< Reference frame rotation
+  Matrix3s oRf_inv_;                                                      //!< Inverse reference rotation
+  boost::shared_ptr<typename StateMultibody::PinocchioModel> pin_model_;  //!< Pinocchio model
+};
+
+template <typename _Scalar>
+struct ResidualDataFrameRotationTpl : public ResidualDataAbstractTpl<_Scalar> {
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  typedef _Scalar Scalar;
+  typedef MathBaseTpl<Scalar> MathBase;
+  typedef ResidualDataAbstractTpl<Scalar> Base;
+  typedef DataCollectorAbstractTpl<Scalar> DataCollectorAbstract;
+  typedef typename MathBase::Vector3s Vector3s;
+  typedef typename MathBase::Matrix3s Matrix3s;
+  typedef typename MathBase::Matrix3xs Matrix3xs;
+  typedef typename MathBase::Matrix6xs Matrix6xs;
+
+  template <template <typename Scalar> class Model>
+  ResidualDataFrameRotationTpl(Model<Scalar>* const model, DataCollectorAbstract* const data)
+      : Base(model, data), rJf(3, 3), fJf(6, model->get_state()->get_nv()) {
+    r.setZero();
+    rRf.setIdentity();
+    rJf.setZero();
+    fJf.setZero();
+    // Check that proper shared data has been passed
+    DataCollectorMultibodyTpl<Scalar>* d = dynamic_cast<DataCollectorMultibodyTpl<Scalar>*>(shared);
+    if (d == NULL) {
+      throw_pretty("Invalid argument: the shared data should be derived from DataCollectorMultibody");
+    }
+
+    // Avoids data casting at runtime
+    pinocchio = d->pinocchio;
+  }
+
+  pinocchio::DataTpl<Scalar>* pinocchio;  //!< Pinocchio data
+  Vector3s r;
```

Comment:
Same here. Why are we redefining r here?

Reply:
Same explanation (see https://github.com/loco-3d/crocoddyl/pull/899#discussion_r604258217)

Reply:
solved